### DISCUSSION TITLE
How to use the redirect-function?

### DISCUSSION THREAD

**Discussion by vaxul** [vercel/next.js#53845] [Help]
### Summary

What I don't understand. Is the example in the documentation of [Functions: redirect](https://nextjs.org/docs/app/api-reference/functions/redirect#example) a good or a bad practice?

I have searched for a similar case like in the example. But following the example result in an `NEXT_REDIRECT` error.

Like described in the text:
> Invoking the `redirect()` function throws a `NEXT_REDIRECT` error and terminates rendering of the route segment in which it was thrown.

### My case
If a specific page-route is called I'd like to check of the data is complete, if not, redirect to a specific page.

Next.js v13.3.4

**Comment by buzzysin** [comment]
I'd also like to hear more about this. I'm currently trying to redirect after submitting a form.

The following describes a client component that handles a form (Formik uses useContext under the hood). I would like to redirect after submission:

```tsx
import { redirect } from "next/navigation"
// ... and others

<Formik<LoginFields>
  initialValues={{ ...loginInitialValues, nextUrl }}
  validationSchema={loginValidator}
  onSubmit={async (values: LoginFields) => {
    try {
      if (values.nextUrl !== undefined && values.nextUrl.length > 0) {
        console.log("Redirecting to: ", values.nextUrl);
        redirect(values.nextUrl);
      } else {
        console.log("Redirecting to: /home");
        redirect("/home");
      }
    } catch (error) {
      if (isRedirectError(error)) {
        // This is where I catch the NEXT_REDIRECT error
      } else {
        throw error;
      }
    }
  }}
>
  {/* ... */}
</Formik>
```

This doesn't work, unless I replace `redirect(...)` with `router.push(...)`, which provides the correct (?) behaviour. However, the docs (and [this article](<https://codevoweb.com/setup-and-use-nextauth-in-nextjs-13-app-directory/#clientside-route-protection>)) seem to indicate that `redirect` can be used this way. But I'm not entirely sure if I am using it correctly, so some clarification would be nice.

**Comment by icyJoseph** [comment]
Hi,

As far as I know, both `notFound` and `redirect`, have to be called while rendering. `While rendering` means in the main body of the function component. Event handlers, use-effects, and in general interactivity, won't really work out. There's no limitation to whether it is a Server, or Client, component being rendered.

The reason is that when you throw as you render, by calling `notFound` or `redirect`, I believe the framework does a check, similar to yours, for the error `instanceof`, and uses that to Route as needed. When you throw inside a client side handler, the render process can't access that error. You should just use router's push/replace for handlers.

Also, you seem to be catching the error, which defeats its purpose anyway.

- For `redirect` called while rendering on client components: https://github.com/vercel/next.js/blob/261db496f7cf04b70f078a6eae0c8baf6fe5c238/packages/next/src/client/components/redirect-boundary.tsx#L52C1-L60
- For `redirect` called while rendering on server components: https://github.com/vercel/next.js/blob/canary/packages/next/src/server/app-render/app-render.tsx#L1490-L1513

And there's more use cases of course.

A client component might redirect while rendering like so:

```tsx
"use client";
import { useState } from "react";
import { redirect } from "next/navigation";

const Flag = ({ flag }: { flag: boolean }) => {
  if (flag) {
    redirect("/about");
  }
  return <div>Not set</div>;
};

export default function Home() {
  const [flag, setFlag] = useState(false);

  return (
    <>
      <Flag flag={flag} />
      <button
        onClick={() => {
          setFlag((x) => !x);
        }}
      >
        set!
      </button>
    </>
  );
}
```

So long story short, if you want the redirect function to do its thing, it has to happen so that the error it throws, is caught by the error boundary.
